Preserve manifest list digest when embedding containers

achilleas-k opened the pull request:
### Skopeo source

This changes the intermediate format of a container **source** from `docker-archive` to `dir`, a directory format that's not yet standardised, but compatibility guidelines are currently being discussed (https://github.com/containers/image/issues/1874).

The `dir` format lets us merge in a multi-image manifest with a container so that we can preserve the manifest digest that was used to specify the image.  This lets users run the image using the same digest that they used to specify it for pulling.  See https://github.com/containers/skopeo/issues/1935 for details on the merge method.

### Skopeo-index source

New source for specifying the manifest-list to download.  This can then be used as a secondary (optional) input to the skopeo stage (see below).  The output of the source is a single file, the manifest that the digest points to.  If the digest points to any other type of object, the source fails.

### Skopeo stage

The skopeo stage now takes an optional list of file references as input.  These files must be manifest-lists.  Each manifest-list is automatically paired with an image source based on the single-image manifest digest.

### Example

Using the [Fedora 37 image on Quay.io](https://quay.io/repository/fedora/fedora/manifest/sha256:d61102cd2dfcb5ac29f752554c8738631245acefb02cae741562349a471fc4d3) as an example, we can now specify the following sources:

```json
    "org.osbuild.skopeo": {
      "items": {
        "sha256:38b8a6c320f32c32d41046e62496fdde0c712a6491a5327146fa7189a23399ad": {
          "image": {
            "name": "quay.io/fedora/fedora",
            "digest": "sha256:b00bf6a9f41f2245ad0c7fccc3a948dbb6266ae808b3f9a8977e9634b3aabdcc"
          }
        }
      }
    },
    "org.osbuild.skopeo-index": {
      "items": {
        "sha256:d61102cd2dfcb5ac29f752554c8738631245acefb02cae741562349a471fc4d3": {
          "image": {
            "name": "quay.io/fedora/fedora"
          }
        }
      }
    }
```

- `sha256:d61102cd2dfcb5ac29f752554c8738631245acefb02cae741562349a471fc4d3`: Manifest-list digest.  This is the multi-image manifest that (in this case) refers to four single-image manifests, one for each architecture.  This is used to download the manifest-list as a single file by the `skopeo-index` stage.
- `sha256:b00bf6a9f41f2245ad0c7fccc3a948dbb6266ae808b3f9a8977e9634b3aabdcc`: Image manifest digest.  This is the single-image manifest for the `amd64` architecture.  This is used to download the image by the `skopeo` stage.
- `sha256:38b8a6c320f32c32d41046e62496fdde0c712a6491a5327146fa7189a23399ad`: Image config digest (the image ID).  This is the ID that appears in `podman image ls` for the container.  This is the ID that's used to store the container in the local osbuild cache and the object ref for the container in the manifest.

The skopeo stage will then look like this:
```json
{
  "type": "org.osbuild.skopeo",
  "inputs": {
    "images": {
      "type": "org.osbuild.containers",
      "origin": "org.osbuild.source",
      "references": {
        "sha256:38b8a6c320f32c32d41046e62496fdde0c712a6491a5327146fa7189a23399ad": {
          "name": "quay.io/fedora/fedora:with-digest"
        }
      }
    },
    "manifest-lists": {
      "type": "org.osbuild.files",
      "origin": "org.osbuild.source",
      "references": [
        "sha256:d61102cd2dfcb5ac29f752554c8738631245acefb02cae741562349a471fc4d3"
      ]
    }
  },
  "options": {
    ...
  }
},
```

The manifest list is parsed for single-image manifest digests and the container is inspected to get its digest.  If no matches are found between the two, the stage raises an exception.  `manifest-lists` are not mandatory, but if they are specified, they should match a container in the `images`.

### Testing

I've tested the above example manually and also tested having multiple containers with matching manifest-lists.
I intend to add tests for the expected behaviour in osbuild-composer when we add support there.

----------

<details>
  <summary>Old summary. No longer relevant but kept for posterity.</summary>

This PR adds support for writing image names to the `names` array of the container store for images that are imported through skopeo.

**Problem**
* Users want to add a container to the blueprint as `<url>@<digest>` and then run the container after build with `podman run <url>@<digest>`
* The way we embed containers in images is by first pulling the container into a docker archive `skopeo copy <url>@<digest> docker-archive:<container>.tar` during the source download phase of the build and then we put it in the container store `skopeo copy docker-archive:<container>.tar <containerstore>:<url>`
* Launching the container with `podman run <url>@<digest>` after booting the OS doesn't work in many cases because the `<digest>` is likely to change.  This is especially the case when relying on the manifest-list digest (which Quay and the Red Hat catalogue present to users).  This is a digest of a manifest that lists multiple container images (usually the same container for different architectures).  When pulling a container using this digest, only the container for the host architecture is pulled.  The local manifest is then completely different from the one used to query the registry.

**Observations**
* When I do a `podman pull <url>@<digest>`, I see that podman puts the digest in the `overlay-images/images.json` as an "alias" for this container image.
* Something similar happens when I copy a container directly into the store with `skopeo copy <url>@<digest> <containerstore>:<url>`.

**Solution**
* Modify our `skopeo` stage (which does `skopeo copy`) to amend the image store `images.json` and add the digest that was used in the blueprint as a name for the container (this PR).
* We can expose this to the user as a `store_source_name` (or something similar) option in the container part of the blueprint (later PR in osbuild-composer).

**Questions/concerns**
* It's important to note that this digest is not the hash of anything that's on disk.  Users want to use this digest because the registries present it as something that uniquely identifies an image, but it should be made clear that the metadata that makes up this digest only exists on the server where this container came from.
    * From what I understand, neither skopeo nor podman actually store any of this metadata.  The name they put in the `images.json` is just that, a name, a string that was used at some point to refer to an image that eventually resolved to the one in the store.
* I've read through the container library sources to make sure that modifying the `images.json` file directly is safe and doesn't need any similar modifications in other places.  I tested this a few times and it seems to simply work, but I'd like a sanity check from some of the container developers at the same time to make sure.
* I've seen a few places where this digest can be stored, depending on how the container is pulled.  If the container is already in the store (under a different name, maybe a tag) and then it's pulled again with the digest, the digest ends up in the `names` array (and the `names-history` at the same time).  If downloaded with `skopeo copy` directly into the store, it's stored as the `digest` of the image.  I chose to put it in `names`, that feels less direct, but I'm open to suggestions.
    * I'm not sure if the names should also be included in the `names-history` array at the same time.


I've asked @runcom for some guidance here.  I've looked at https://github.com/containers/image/commit/e1ee980e6d26ce30d123a61f4a91ee1347c4d67c and the linked issues, but that doesn't work because we rely on docker-archive for the local intermediate storage.  That's one thing we can revisit in the future, whether a different local intermediate storage might make things easier, without having any negative side effects.

</details>

runcom commented:
cc @mtrmac @mheon - could give us some more guidance/check on what we're doing here? :angel:

mtrmac reviewed (changes requested):
TL:DR; This approach is unworkable and must be rethought.

---

> * The way we embed containers in images is by first pulling the container into a docker archive `skopeo copy <url>@<digest> docker-archive:<container>.tar` during the source download phase of the build and then we put it in the container store `skopeo copy docker-archive:<container>.tar <containerstore>:<url>`

If at all possible, don’t do that. `docker-archive:`, besides being slow, _always_ loses information. Use `dir:` if you absolutely need a non-registry file store, and want to preserve the image as it is on in-registry.

> * Modify our `skopeo` stage (which does `skopeo copy`) to amend the image store `images.json` 

_Definitely_ don’t do that in production. The format of `images.json` is an internal implementation detail and may change at any time (in fact, _this month_ we have seriously discussed what it would take to move to SQLite).

The _locking_ of `images.json` is even more of an internal implementation detail, and critical to reliability. This code does not do any locking, and may therefore interfere with any running Podman/Skopeo/… processes in unanticipated ways.

The right way to add a name to an image is `podman tag`, or, I suppose, `skopeo copy containers-storage:$old containers-storage:$new`.

> and add the digest that was used in the blueprint as a name for the container (this PR).

That won’t work anyway, because readers of those images will need to actually find the manifest _data_ that matches that digest. Just adding a name does not record the manifest data.

So, the entire premise of using `docker-archive` should be revisited. That’s the easiest way to get this done. (An alternative would be a custom-written tool to download the right manifests, and a custom-written Go tool to add the manifests to the store. Let’s not.)

> * It's important to note that this digest is not the hash of anything that's on disk.
> …
>   * From what I understand, neither skopeo nor podman actually store any of this metadata.

This is incorrect. AFAICS all digest reference accesses like that will fail.

> * I've read through the container library sources to make sure that modifying the `images.json` file directly is safe

It isn’t safe. It’s not up to me but speaking for myself, I would _absolutely_ refuse to support any modifications not implemented by calling a (reasonably current) version of containers/storage.


---

The implementation is also not ensuring that each name is only assigned to only one image.

achilleas-k commented:
> TL:DR; This approach is unworkable and must be rethought.
> 
> > * The way we embed containers in images is by first pulling the container into a docker archive `skopeo copy <url>@<digest> docker-archive:<container>.tar` during the source download phase of the build and then we put it in the container store `skopeo copy docker-archive:<container>.tar <containerstore>:<url>`
> 
> If at all possible, don’t do that. `docker-archive:`, besides being slow, _always_ loses information. Use `dir:` if you absolutely need a non-registry file store, and want to preserve the image as it is on in-registry.
> 

Thanks.  I think this is probably the best way forward and probably wont require too much reworking.

> > * Modify our `skopeo` stage (which does `skopeo copy`) to amend the image store `images.json`
> 
> _Definitely_ don’t do that in production. The format of `images.json` is an internal implementation detail and may change at any time (in fact, _this month_ we have seriously discussed what it would take to move to SQLite).
> 

I see.  That's great to know.  It wouldn't be a problem to rely on a shell utility that can _properly_ write to the store, it would have to be something that's packaged for RHEL.

> The _locking_ of `images.json` is even more of an internal implementation detail, and critical to reliability. This code does not do any locking, and may therefore interfere with any running Podman/Skopeo/… processes in unanticipated ways.
> 

This is not an issue in our case (the stages are run in a clean bubblewrap container and no other processes should be running), but I understand the concern.  Thank you.

> The right way to add a name to an image is `podman tag`, or, I suppose, `skopeo copy containers-storage:$old containers-storage:$new`.
> 

We need an intermediate local storage because the download stage happens separately from the build stage.  That said, if a `dir:` or an intermediate `container-store:` gets us around this issue, we can work with that.

> > and add the digest that was used in the blueprint as a name for the container (this PR).
> 
> That won’t work anyway, because readers of those images will need to actually find the manifest _data_ that matches that digest. Just adding a name does not record the manifest data.
> 

From some testing I did, I found that this does work (adding the name) and no manifest data was pulled (e.g., when doing `podman run <name>@<manifest digest>`), but perhaps this is an unintended consequence or side effect?
I couldn't find any information on where or how the manifest or manifest-list is stored.  Perhaps it's not important if we do this properly (with `dir` or `container-store`), but it'd be good to know.

> So, the entire premise of using `docker-archive` should be revisited. That’s the easiest way to get this done. (An alternative would be a custom-written tool to download the right manifests, and a custom-written Go tool to add the manifests to the store. Let’s not.)
> 
> > * It's important to note that this digest is not the hash of anything that's on disk.
> >   …
> > * From what I understand, neither skopeo nor podman actually store any of this metadata.
> 
> This is incorrect. AFAICS all digest reference accesses like that will fail.
> 
> > * I've read through the container library sources to make sure that modifying the `images.json` file directly is safe
> 
> It isn’t safe. It’s not up to me but speaking for myself, I would _absolutely_ refuse to support any modifications not implemented by calling a (reasonably current) version of containers/storage.
> 

Good to know.

> The implementation is also not ensuring that each name is only assigned to only one image.

You're right, yes, this is an oversight.

achilleas-k commented:
Adding some info after speaking with @mtrmac:

> > That won’t work anyway, because readers of those images will need to actually find the manifest _data_ that matches that digest. Just adding a name does not record the manifest data.
> 
> From some testing I did, I found that this does work (adding the name) and no manifest data was pulled (e.g., when doing `podman run <name>@<manifest digest>`), but perhaps this is an unintended consequence or side effect? I couldn't find any information on where or how the manifest or manifest-list is stored. Perhaps it's not important if we do this properly (with `dir` or `container-store`), but it'd be good to know.
> 

They're in the image's directory and their digests are in the `big-data-digests` array (example below).
```
# jq '.[]."big-data-digests"' storage/overlay-images/images.json
{
  "manifest": "sha256:4db63516a5990cc27f1655366bf5ea2fbde97ea6295b028f916bfc6de7d96d4c",
  "manifest-sha256:4db63516a5990cc27f1655366bf5ea2fbde97ea6295b028f916bfc6de7d96d4c": "sha256:4db63516a5990cc27f1655366bf5ea2fbde97ea6295b028f916bfc6de7d96d4c",
  "sha256:7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6": "sha256:7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6",
  "signature-4db63516a5990cc27f1655366bf5ea2fbde97ea6295b028f916bfc6de7d96d4c": "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
}
# sha256sum storage/overlay-images/7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6/*
4db63516a5990cc27f1655366bf5ea2fbde97ea6295b028f916bfc6de7d96d4c  storage/overlay-images/7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6/=bWFuaWZlc3Qtc2hhMjU2OjRkYjYzNTE2YTU5OTBjYzI3ZjE2NTUzNjZiZjVlYTJmYmRlOTdlYTYyOTViMDI4ZjkxNmJmYzZkZTdkOTZkNGM=
7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6  storage/overlay-images/7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6/=c2hhMjU2OjdhY2Q5ZTU1ZTg3NTJjOGE2NGU4NTY5ZWEzZTI3MzZjZWI4OWJiNGQwOTliYWU1OTdkMjAyZWY4ODJlMTAzZTY=
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  storage/overlay-images/7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6/=c2lnbmF0dXJlLTRkYjYzNTE2YTU5OTBjYzI3ZjE2NTUzNjZiZjVlYTJmYmRlOTdlYTYyOTViMDI4ZjkxNmJmYzZkZTdkOTZkNGM=
4db63516a5990cc27f1655366bf5ea2fbde97ea6295b028f916bfc6de7d96d4c  storage/overlay-images/7acd9e55e8752c8a64e8569ea3e2736ceb89bb4d099bae597d202ef882e103e6/manifest
```

It _shouldn't_ work without the data being there when the name contains a digest but it does.  This is a bug.

mtrmac commented:
> > The right way to add a name to an image is `podman tag`, or, I suppose, `skopeo copy containers-storage:$old containers-storage:$new`.
> 
> We need an intermediate local storage because the download stage happens separately from the build stage.

(I was thinking this could be done in the build stage; that’s just a way to add a tag without writing an external utility, e.g. with the `skopeo copy`, $old and $new would use the same _store_ but just different image names. But that should work for names with _tags_, it gets complicated with digests.)

mtrmac commented:
> It _shouldn't_ work without the data being there when the name contains a digest but it does. This is a bug.

https://github.com/containers/podman/issues/17682

achilleas-k commented:
Everything changed.  I've updated the PR description based on the solutions discussed with the containers team.

mtrmac commented on sources/org.osbuild.skopeo:
```diff
@@ -75,25 +81,27 @@ class SkopeoSource(sources.SourceService):
 
-            # The archive format can't store signatures, but we still verify them during download
-            extra_args.append("--remove-signatures")
+            if archive_format == "docker-archive":
```
BTW `oci-archive` does not support signatures either.

mtrmac commented on sources/org.osbuild.skopeo:
```diff
@@ -103,17 +111,43 @@ class SkopeoSource(sources.SourceService):
+            # calculate the checksum of the manifest of the container image in the destination
+            manifest_path = os.path.join(destination, "manifest.json")
+            with open(manifest_path, encoding="utf-8") as manifest:
+                manifest_checksum = hashlib.sha256(manifest.read().encode()).hexdigest()
```
Mostly pedantically, this should use `skopeo manifest-digest` instead of raw SHA256 to compute the right digest for schema1 images.

OTOH schema1 images can’t be multi-arch, so this would only result in a different manifest path for a file that would be ignored (just like this code creates `*.manifest.json` files that are ignored for other non-multiarch images).

achilleas-k replied:
Still, best to use the tooling when it exists.

mtrmac reviewed (commented):
(I’m unfamiliar with the code base, this is just a quick skim of the container image-related aspects.)

thozza reviewed (commented):
This looks great and pretty clean 👍  It is great that you figured out a reasonable solution for this 🐛 .  The only thing that I'm not sure about is the ability to specify the format for a stage input. I think that it just add new options to shoot oneself in the foot for not obvious (to me) benefit.

mtrmac commented on inputs/org.osbuild.containers:
```diff
@@ -32,6 +34,12 @@ SCHEMA = r"""
         "type": "string",
         "description": "The name to use for the image"
+      },
+      "format": {
+        "type": "string",
+        "enum": ["docker-archive", "oci-archive", "dir"],
+        "description": "The storage format of the container archive",
+        "default": "docker-archive"
```
Specifically `docker-archive` is not a “source” format, in the sense that it’s not just a serialization of the registry data (that’s actually mostly what `dir:` is). Using `docker-archive` implies a _lossy_ conversion (decompress layers, drop the manifest; can’t represent multiple multi-arch variants).

(As an outsider, it also seems attractive to me to just hard-code one format instead of having to test and support three — but I wouldn’t know if there were any compatibility constraints e.g. with a version skew between consumers and producers.)

thozza replied:
I meant "source" in the **osbuild** terminology. Meaning a source that can be referenced by osbuild stage as an input. I used "source format" in this case to refer to the value of the "format" property of an osbuild stage input of type `org.osbuild.containers` referenced from the `source` section of an osbuild manifest.

The point I'm trying to make is that IMO, the producer (a source definition using a specific "format" for storing the container or an osbuild pipeline producing a container stored using a specific "format") predetermines the most sensible "format" to use for a stage  `org.osbuild.containers` input. I'm not saying that we should hardcode one specific "format". There is IMO just no point in allowing for it to be explicitly specified for the stage input, instead I suggest to make it implicit.

If I understand the PR and all comments correctly, it should be possible mostly to loose some information by specifying `docker-archive` or `oci-archive` as the format in a stage input, if the container was stored using `dir` format.

Therefore defaulting to the "format" in a stage `org.osbuild.containers` input that is used to store the container seems logical to me.

This does not imply that we can't test all "formats".

achilleas-k commented:
> The only thing that I'm not sure about is the ability to specify the format for a stage input. I think that it just add new options to shoot oneself in the foot for not obvious (to me) benefit.

You're not wrong.  It's a bit strange and creates a situation where the stage input needs to match the source.  In fact, it's even weirder than that because you can specify the stage input to have a different format than the source, but if the input format was already downloaded and is in the store, it would work.

My thinking on this was that I didn't want to introduce a behaviour change without introducing an option.  On the other hand, this is an internal implementation detail and it might not matter.  You're right that the most powerful format right now is the `dir`, so we could use that everywhere, or at least only use it for sources and keep pipeline artifacts as `oci-archive` to minimise changes.  My one issue with this that the `dir` format isn't standardised and I intend to use it only when this new behaviour is desired, keeping the old behaviour intact for the default case.  But perhaps this is the wrong way to think about it and we could go all in with `dir` and bring back other formats if we ever run into incompatibilities.

I thought I was making the most cautious choice here, but you're making me reconsider now.

achilleas-k commented:
Rewrote with `dir` replacing `docker-archive` and dropped the options.

achilleas-k commented:
Updated the original comment again, though the Fedora digests are outdated now.

achilleas-k commented:
Fixed two small issues caught by pylint.

achilleas-k commented on sources/org.osbuild.skopeo:
```diff
@@ -103,17 +98,43 @@ class SkopeoSource(sources.SourceService):
 
-            # Atomically move download dir into place on successful download
+            # fetch the manifest and merge it into the archive
+            self.merge_manifest(source, dest_path, extra_args)
```
This function should only run when the digest points to a manifest list.  Otherwise, we're doing extra work for no reason.  Running `skopeo copy --multi-arch=index-only` with a single image digest will pull the image, so we end up copying the same archive twice and duplicating the manifest in the merged directory.  Is that right @mtrmac?

mtrmac replied:
Ugh, yes, `--multi-arch=index-only` with a non-multi-arch image currently copies everything. It’s clearly documented that way, but it’s also rather unhelpful for this use case. (OTOH the fix would be to _fail_ with non-multi-arch, which I suppose, doesn’t help :) )

Either way, that leads to a question of “how does a caller tell if an image is multi-arch”, and the _good_ answer to that is an unimplemented RFE https://github.com/containers/skopeo/issues/1283  . Without that, it’s possible, although not at all elegant, to decide that outside of Skopeo using `inspect --raw` and inspecting that JSON manually.

achilleas-k replied:
> Without that, it’s possible, although not at all elegant, to decide that outside of Skopeo using inspect --raw and inspecting that JSON manually.

Yeah that was my plan now.  Inspect it and look at the type.

achilleas-k commented:
Made the changes we discussed today.  Manifest digest is now stored separately in the cache and it is only merged when the input option requests it.

achilleas-k commented:
Fixed a bug, added docstrings, and changed the name of the new option from `manifest_digest` to `list_digest` to make it clear that it must be the digest of a manifest list.

achilleas-k commented on sources/org.osbuild.skopeo:
```diff
@@ -79,22 +79,15 @@ class SkopeoSource(sources.SourceService):
-            destination = f"docker-archive:{archive_path}"
+            # We use the dir format because it is the most powerful in terms of feature support and is the closest to a
+            # direct serialisation of the registry data.
+            dir_name = "image"
```
Good idea.

mtrmac commented:
> @mtrmac is it possible for the `dir` format to have subdirectories?

It doesn’t happen currently … but compare all the uncertainty in https://github.com/containers/image/issues/1874 .

achilleas-k commented:
> > @mtrmac is it possible for the `dir` format to have subdirectories?
> 
> It doesn’t happen currently … but compare all the uncertainty in [containers/image#1874](https://github.com/containers/image/issues/1874) .

Understood.  Thanks.

achilleas-k commented:
Let's hope this is the last one :)

achilleas-k commented:
Updated PR description.

supakeen commented on sources/org.osbuild.skopeo:
```diff
@@ -103,17 +107,21 @@ class SkopeoSource(sources.SourceService):
             with ctx.suppress_oserror(errno.ENOTEMPTY, errno.EEXIST):
-                os.rename(archive_dir, f"{self.cache}/{image_id}")
+                os.makedirs(f"{self.cache}/{image_id}", exist_ok=True)
+                os.rename(f"{archive_dir}/{self.dir_name}", f"{self.cache}/{image_id}/{self.dir_name}")
```
Please use path concatenation functions on the above two lines instead of string formatting, either `os.path.join` or `pathlib` would be fine.

supakeen commented on stages/org.osbuild.skopeo:
```diff
@@ -53,49 +59,127 @@ SCHEMA_2 = r"""
+                source = f"dir:{tmp_source}"
             elif container_format == "oci-archive":
-                source = f"oci-archive:{linkname}"
+                source = f"oci-archive:{tmp_source}"
```
```
if container_format not in ("dir", "docker-archive"):
    raise RuntimeError(...)

source = f"{container_format}:{tmp_source}"
```

supakeen reviewed (commented):
Just two minor things Python-wise.

achilleas-k commented:
> Just two minor things Python-wise.

Thanks!  Fixed.

thozza reviewed (dismissed):
Looks great. The only thing that I'm wondering about is if it would be possible to amend the `skopeo` stage unit test to test the manifest merging 🤔

achilleas-k commented:
> Looks great. The only thing that I'm wondering about is if it would be possible to amend the `skopeo` stage unit test to test the manifest merging thinking

I thought about this a lot and there's definitely something we can do here but it'll need some setting up.  The manifest merging only works for containers through sources now and we only unit test container embedding through a pipeline, with a tiny container as an inline source.  I think the best way to test this would be in composer with a container we make ourselves with a manifest list (could include both x86 and arm64), possibly hosted on quay or the gitlab registry.

That all said, there's no reason we can't add the same container to a test here, but I'm not sure how we would verify the manifest merging (given how the stage unit tests run).  It would at least be a good way to test that there are no build-breaking bugs.

mtrmac commented:
> That all said, there's no reason we can't add the same container to a test here, but I'm not sure how we would verify the manifest merging (given how the stage unit tests run).

(Local unit-test: verify the directory has exactly the same contents.) To test integration with Skopeo, possibly `skopeo inspect --raw dir:…` to verify this reports the multi-arch manifest; and `skopeo inspect` to verify this shows the right config etc. of the per-arch instance.

achilleas-k commented:
Added a test to the skopeo stage for the new feature.  See the last two commits.  The other commits didn't change, but were rebased on `main`).

For the test, I created a simple container with a manifest list and pushed it to the osbuild-composer registry on gitlab [1].  I plan to use the same one for osbuild-composer tests.  I put the creation instructions in the commit message.

[1] https://gitlab.com/redhat/services/products/image-builder/ci/osbuild-composer/container_registry/4039574

thozza reviewed (approved):
Great work (on figuring our everything and also the test case), I love it 😍

achilleas-k commented:
@thozza Waiting for one more pair of eyes before merging?  Maybe @supakeen?

thozza commented:
> @thozza Waiting for one more pair of eyes before merging? Maybe @supakeen?

Yes, I wanted to let @supakeen have a chance to review. If he is not going to do it, we can merge 😉
